src/net.rs

```diff
@@ -672,8 +671,7 @@ unsafe fn connect_overlapped(
     };
 
     let ptr = CONNECTEX.get(socket)?;
-    assert!(ptr != 0);
-    let connect_ex = mem::transmute::<_, LPFN_CONNECTEX>(ptr);
+    let connect_ex = mem::transmute::<_, LPFN_CONNECTEX>(ptr).unwrap();
```

Comment:
`LPFN_CONNECTEX` etc. have `extern "system" fn` type:

https://docs.rs/windows-sys/latest/windows_sys/Win32/Networking/WinSock/type.LPFN_CONNECTEX.html

`T` has the same size as `Option<T>` and one can `mem::transmute` from all valid values of `T` to `Option<T>`, if `T` is `extern "system" fn`:

https://doc.rust-lang.org/stable/std/option/index.html#representation
https://doc.rust-lang.org/stable/std/option/index.html#fn1

It does say though that this doesn't work for `None` values. This code previously asserted that the pointer values were non-zero, so this shouldn't ever happen in practice. To be safe though, I've now reinstated those asserts to prevent this form of UB.

The other question is whether `usize` has the same representation as `extern "system" fn`. I don't know where that's documented, but the code was already making that assumption.

Reply:
> one can mem::transmute from all valid values of `T` to `Option<T>`

I don't think zero is a valid `extern "system" fn` pointer though. The point is that the memory layout of `Option<T>` is equal to `T` for all *valid* `T`s. That's what the niche optimization is about. But the compiler is free to pick any representation for the `None` value out of all possible invalid representations for `T`. And we should not implicitly rely on zero being the only one or the one it will pick.

Great that you added the asserts back. Unless we can prove it's safe without them, I much rather prefer this.

Reply:
> one can mem::transmute from all valid values of `T` to `Option<T>`

I don't think zero is a valid `extern "system" fn` pointer though. The point is that the memory layout of `Option<T>` is equal to `T` for all *valid* `T`s. That's what the niche optimization is about. But the compiler is free to pick any representation for the `None` value out of all possible invalid representations for `T`. And we should not implicitly rely on zero being the only one or the one it will pick.

Great that you added the asserts back. Unless we can prove it's safe without them, I much rather prefer this.